### PR TITLE
Allow json_decode in parse_params to be optional

### PR DESCRIPTION
**Pull request by eiri**
This allows explicitly indicate if function `parse_params` should decode passed parameter's value instead of trying to guess it by value's term type.

It makes it possible to use this function both with GET and POST queries.

COUCHDB-3031

### CODE REVIEW

File: src/couch_mrview_http.erl [hunk 1]
```diff
@@ -439,39 +440,43 @@ parse_params(Props, Keys) ->
...
+    parse_param(binary_to_list(Key), Val, Args, IsDecoded);
+parse_param(Key, Val, Args, IsDecoded) ->
     case Key of
         "" ->
             Args;
         "reduce" ->
             Args#mrargs{reduce=parse_boolean(Val)};
         "key" ->
-            JsonKey = parse_json(Val),
+            JsonKey = parse_json(Val, IsDecoded),
```
Comment (rnewson) [inline src/couch_mrview_http.erl]: now that we look again, it's odd to have a function called parse_json that conditionally doesn't parse json. How about a clause for each right here?;

-parse_param(Key, Val, Args) when is_binary(Key) ->
-    parse_param(binary_to_list(Key), Val, Args);
-parse_param(Key, Val, Args) ->
+parse_param(Key, Val, Args, IsDecoded) when is_binary(Key) ->
+    parse_param(binary_to_list(Key), Val, Args, IsDecoded);
+parse_param(Key, Val, Args, IsDecoded) ->
     case Key of
         "" ->
             Args;
         "reduce" ->
             Args#mrargs{reduce=parse_boolean(Val)};
+        "key" when IsDecoded ->
+            Args#mrargs{start_key=Val, end_key=Val};
         "key" ->
             JsonKey = ?JSON_DECODE(Val),
             Args#mrargs{start_key=JsonKey, end_key=JsonKey};

**Comment by davisp** [comment]
+1

**Comment by eiri** [comment]
@rnewson Changes it to use your suggestion.

**Comment by rnewson** [comment]
+1

**Pull request merged**